fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java

```diff
@@ -0,0 +1,405 @@
+        this.rwLock.writeLock().unlock();
+    }
+
+    public Warehouse(long id, String name) {
+        this.id = id;
+        this.name = name;
+        this.rwLock = new QueryableReentrantReadWriteLock(true);
+    }
+
+    public Warehouse(long id, String name, Map<String, String> properties) {
+        this(id, name);
+        if (properties != null) {
+            if (properties.get("size") != null) {
+                size = properties.get("size");
+            }
+
+            if (properties.get("min_cluster") != null) {
+                minCluster = Integer.valueOf(properties.get("min_cluster"));
+            }
+
+            if (properties.get("max_cluster") != null) {
+                maxCluster = Integer.valueOf(properties.get("max_cluster"));
+            }
+        }
+    }
+
+    public Warehouse() {
+        this(0, null);
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    public String getFullName() {
+        return name;
+    }
+
+    public void setState(WarehouseState state) {
+        this.state = state;
+    }
+
+    public WarehouseState getState() {
+        return state;
+    }
+    // property setters and getters
+    public void setSize(String size) {
+        this.size = size;
+    }
+    public String getSize() {
+        return size;
+    }
+
+    public void setMinCluster(int minCluster) {
+        this.minCluster = minCluster;
+    }
+
+    public int getMinCluster() {
+        return minCluster;
+    }
+
+    public void setMaxCluster(int maxCluster) {
+        this.maxCluster = maxCluster;
+    }
+
+    public int getMaxCluster() {
+        return maxCluster;
+    }
+
+    // and others...
+
+    public long getTotalRunningSqls() {
+        return -1L;
+    }
+
+    public int getTotalPendingSqls() {
+        return -1;
+    }
+
+    public void setPendingSqls(int val) {
+    }
+
+    public int addAndGetPendingSqls(int delta) {
+        return 1;
+    }
+
+    public void setClusters(Map<Long, Cluster> clusters) {
+        this.clusters = clusters;
+    }
+
+    public Map<Long, Cluster> getClusters() {
+        return clusters;
+    }
+
+    public void setExist(boolean exist) {
+        this.exist = exist;
+    }
+
+    public List<List<String>> getClustersInfo() {
+        return procNode.fetchResult().getRows();
+    }
+
+    public void suspendSelf(boolean isReplay) throws DdlException {
+        writeLock();
+        try {
+            this.state = WarehouseState.SUSPENDED;
+            if (!isReplay) {
+                releaseComputeNodes();
+            }
+            clusters.clear();
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void resumeSelf() throws DdlException {
+        writeLock();
+        try {
+            clusters = applyComputeNodes();
+            this.state = WarehouseState.RUNNING;
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void dropSelf() throws DdlException {
+        readLock();
```

Comment:
we need write lock here?

Reply:
In my opinion, read lock is enough, because we won't change the state or clusters of this warehouse.